File: solutions/python/trees/maximum-difference-between-node-and-ancestor.py
# LeetCode 1026 — Maximum Difference Between Node and Ancestor (Medium)
# Category: Trees · Approach: Track min/max on path
# Time: O(n) | Space: O(h)
# Source: https://leetcode.com/problems/maximum-difference-between-node-and-ancestor/

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val, self.left, self.right = val, left, right

def max_ancestor_diff(root: TreeNode) -> int:
    best = 0

    def dfs(node, lo, hi):
        nonlocal best
        if not node:
            best = max(best, hi - lo)
            return
        lo = min(lo, node.val)
        hi = max(hi, node.val)
        dfs(node.left, lo, hi)
        dfs(node.right, lo, hi)

    dfs(root, root.val, root.val)
    return best

File: solutions/typescript/trees/maximum-difference-between-node-and-ancestor.ts
// LeetCode 1026 — Maximum Difference Between Node and Ancestor (Medium)
// Category: Trees · Approach: Track min/max on path
// Time: O(n) | Space: O(h)
// Source: https://leetcode.com/problems/maximum-difference-between-node-and-ancestor/

class TreeNode {
  val: number;
  left: TreeNode | null = null;
  right: TreeNode | null = null;
  constructor(val = 0) {
    this.val = val;
  }
}

function maxAncestorDiff(root: TreeNode): number {
  let best = 0;
  const dfs = (node: TreeNode | null, lo: number, hi: number) => {
    if (!node) {
      best = Math.max(best, hi - lo);
      return;
    }
    lo = Math.min(lo, node.val);
    hi = Math.max(hi, node.val);
    dfs(node.left, lo, hi);
    dfs(node.right, lo, hi);
  };
  dfs(root, root.val, root.val);
  return best;
}
